-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33337/21895

ERROR: Build errors found during clang-tidy run.
```
--
FWCore/Framework/src/Worker.h:975:9: error: unknown type name 'ServiceWeakToken'; did you mean 'ServiceToken'? [clang-diagnostic-error]
        ServiceWeakToken weakToken = token;
        ^~~~~~~~~~~~~~~~
        ServiceToken
--
FWCore/Framework/src/Worker.h:979:56: error: no member named 'lock' in 'edm::ServiceToken' [clang-diagnostic-error]
              ServiceRegistry::Operate guard(weakToken.lock());
                                                       ^
FWCore/Framework/src/Worker.h:981:42: error: no member named 'lock' in 'edm::ServiceToken' [clang-diagnostic-error]
                               weakToken.lock(),
                                         ^
FWCore/Framework/src/Worker.h:1046:7: error: unknown type name 'ServiceWeakToken'; did you mean 'ServiceToken'? [clang-diagnostic-error]
      ServiceWeakToken weakToken = serviceToken;
      ^~~~~~~~~~~~~~~~
      ServiceToken
--
FWCore/Framework/src/Worker.h:1052:52: error: no member named 'lock' in 'edm::ServiceToken' [clang-diagnostic-error]
          ServiceRegistry::Operate guard(weakToken.lock());
                                                   ^
Suppressed 1163 warnings (1162 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

--- a/FWCore/Framework/interface/EDLooperBase.h
+++ b/FWCore/Framework/interface/EDLooperBase.h
@@ -54,6 +54,7 @@
 //
 
 #include "DataFormats/Provenance/interface/ModuleDescription.h"
+#include "FWCore/Framework/interface/EDConsumerBase.h"
 #include "FWCore/Framework/interface/Frameworkfwd.h"
 #include "FWCore/ServiceRegistry/interface/ModuleCallingContext.h"
 #include "FWCore/Utilities/interface/propagate_const.h"
@@ -74,12 +75,12 @@ namespace edm {
   class ProcessingController;
   class ActivityRegistry;
 
-  class EDLooperBase {
+  class EDLooperBase : public EDConsumerBase {
   public:
     enum Status { kContinue, kStop };
 
     EDLooperBase();
-    virtual ~EDLooperBase() noexcept(false);
+    ~EDLooperBase() noexcept(false) override;
 
     EDLooperBase(EDLooperBase const&) = delete;             // Disallow copying and moving
     EDLooperBase& operator=(EDLooperBase const&) = delete;  // Disallow copying and moving

--- a/FWCore/Framework/src/EDLooperBase.cc
+++ b/FWCore/Framework/src/EDLooperBase.cc
@@ -52,11 +52,13 @@ namespace edm {
     ParentContext parentContext(streamContext);
     ModuleContextSentry moduleContextSentry(&moduleCallingContext_, parentContext);
     Event event(eventPrincipal, moduleDescription_, &moduleCallingContext_);
+    event.setConsumer(this);
 
     Status status = kContinue;
     try {
       ESParentContext parentC(&moduleCallingContext_);
-      const EventSetup es{esi, static_cast<unsigned int>(Transition::Event), nullptr, parentC, false};
+      const EventSetup es{
+          esi, static_cast<unsigned int>(Transition::Event), esGetTokenIndices(Transition::Event), parentC, false};
       status = duringLoop(event, es, ioController);
     } catch (cms::Exception& e) {
       e.addContext("Calling the 'duringLoop' method of a looper");
@@ -72,7 +74,8 @@ namespace edm {
 
   EDLooperBase::Status EDLooperBase::doEndOfLoop(const edm::EventSetupImpl& esi) {
     ESParentContext parentC(&moduleCallingContext_);
-    const EventSetup es{esi, static_cast<unsigned int>(Transition::EndRun), nullptr, parentC, false};
+    const EventSetup es{
+        esi, static_cast<unsigned int>(Transition::EndRun), esGetTokenIndices(Transition::EndRun), parentC, false};
     return endOfLoop(es, iCounter_);
   }
 
@@ -86,7 +89,11 @@ namespace edm {
 
   void EDLooperBase::beginOfJob(const edm::EventSetupImpl& iImpl) {
     ESParentContext parentC(&moduleCallingContext_);
-    beginOfJob(EventSetup{iImpl, static_cast<unsigned int>(Transition::BeginRun), nullptr, parentC, false});
+    beginOfJob(EventSetup{iImpl,
+                          static_cast<unsigned int>(Transition::BeginRun),
+                          esGetTokenIndices(Transition::BeginRun),
+                          parentC,
+                          false});
   }
   void EDLooperBase::beginOfJob(const edm::EventSetup&) { beginOfJob(); }
   void EDLooperBase::beginOfJob() {}
@@ -103,8 +110,10 @@ namespace edm {
     ParentContext parentContext(&globalContext);
     ModuleContextSentry moduleContextSentry(&moduleCallingContext_, parentContext);
     Run run(iRP, moduleDescription_, &moduleCallingContext_, false);
+    run.setConsumer(this);
     ESParentContext parentC(&moduleCallingContext_);
-    const EventSetup es{iES, static_cast<unsigned int>(Transition::BeginRun), nullptr, parentC, false};
+    const EventSetup es{
+        iES, static_cast<unsigned int>(Transition::BeginRun), esGetTokenIndices(Transition::BeginRun), parentC, false};
     beginRun(run, es);
   }
 
@@ -118,8 +127,10 @@ namespace edm {
     ParentContext parentContext(&globalContext);
     ModuleContextSentry moduleContextSentry(&moduleCallingContext_, parentContext);
     Run run(iRP, moduleDescription_, &moduleCallingContext_, true);
+    run.setConsumer(this);
     ESParentContext parentC(&moduleCallingContext_);
-    const EventSetup es{iES, static_cast<unsigned int>(Transition::EndRun), nullptr, parentC, false};
+    const EventSetup es{
+        iES, static_cast<unsigned int>(Transition::EndRun), esGetTokenIndices(Transition::EndRun), parentC, false};
     endRun(run, es);
   }
   void EDLooperBase::doBeginLuminosityBlock(LuminosityBlockPrincipal& iLB,
@@ -134,8 +145,13 @@ namespace edm {
     ParentContext parentContext(&globalContext);
     ModuleContextSentry moduleContextSentry(&moduleCallingContext_, parentContext);
     LuminosityBlock luminosityBlock(iLB, moduleDescription_, &moduleCallingContext_, false);
+    luminosityBlock.setConsumer(this);
     ESParentContext parentC(&moduleCallingContext_);
-    const EventSetup es{iES, static_cast<unsigned int>(Transition::BeginLuminosityBlock), nullptr, parentC, false};
+    const EventSetup es{iES,
+                        static_cast<unsigned int>(Transition::BeginLuminosityBlock),
+                        esGetTokenIndices(Transition::BeginLuminosityBlock),
+                        parentC,
+                        false};
     beginLuminosityBlock(luminosityBlock, es);
   }
   void EDLooperBase::doEndLuminosityBlock(LuminosityBlockPrincipal& iLB,
@@ -150,8 +166,13 @@ namespace edm {
     ParentContext parentContext(&globalContext);
     ModuleContextSentry moduleContextSentry(&moduleCallingContext_, parentContext);
     LuminosityBlock luminosityBlock(iLB, moduleDescription_, &moduleCallingContext_, true);
+    luminosityBlock.setConsumer(this);
     ESParentContext parentC(&moduleCallingContext_);
-    const EventSetup es{iES, static_cast<unsigned int>(Transition::EndLuminosityBlock), nullptr, parentC, false};
+    const EventSetup es{iES,
+                        static_cast<unsigned int>(Transition::EndLuminosityBlock),
+                        esGetTokenIndices(Transition::EndLuminosityBlock),
+                        parentC,
+                        false};
     endLuminosityBlock(luminosityBlock, es);
   }
 

--- a/FWCore/Framework/src/EventProcessor.cc
+++ b/FWCore/Framework/src/EventProcessor.cc
@@ -641,6 +641,18 @@ namespace edm {
     }
     espController_->finishConfiguration();
     schedule_->beginJob(*preg_, esp_->recordsToProxyIndices());
+    if (looper_) {
+      constexpr bool mustPrefetchMayGet = true;
+      auto const processBlockLookup = preg_->productLookup(InProcess);
+      auto const runLookup = preg_->productLookup(InRun);
+      auto const lumiLookup = preg_->productLookup(InLumi);
+      auto const eventLookup = preg_->productLookup(InEvent);
+      looper_->updateLookup(InProcess, *processBlockLookup, mustPrefetchMayGet);
+      looper_->updateLookup(InRun, *runLookup, mustPrefetchMayGet);
+      looper_->updateLookup(InLumi, *lumiLookup, mustPrefetchMayGet);
+      looper_->updateLookup(InEvent, *eventLookup, mustPrefetchMayGet);
+      looper_->updateLookup(esp_->recordsToProxyIndices());
+    }
     // toerror.succeeded(); // should we add this?
     for_all(subProcesses_, [](auto& subProcess) { subProcess.doBeginJob(); });
     actReg_->postBeginJobSignal_();

--- a/FWCore/Framework/test/BuildFile.xml
+++ b/FWCore/Framework/test/BuildFile.xml
@@ -138,7 +138,7 @@
   <use name="FWCore/Framework"/>
 </library>
 
-<library file="stubs/DummyLooper.cc,stubs/TestModuleChangeLooper.cc" name="FWCoreFrameworkTestDummyLooper">
+<library file="stubs/DummyLooper.cc,stubs/TestModuleChangeLooper.cc,stubs/IntTestLooper.cc" name="FWCoreFrameworkTestDummyLooper">
   <flags EDM_PLUGIN="1"/>
   <lib name="FWCoreFrameworkTestDummyForEventSetup"/>
   <use name="FWCore/Framework"/>

--- a/FWCore/Framework/test/run_esproducerlooper.sh
+++ b/FWCore/Framework/test/run_esproducerlooper.sh
@@ -10,4 +10,4 @@ function die { echo $1: status $2 ;  exit $2; }
 #(cmsRun ${LOCAL_TEST_DIR}/test_esproducerlooper_prefer_not_source_cfg.py ) || die 'Failure using test_esproducerlooper_prefer_not_source_cfg.py' $?
 (cmsRun ${LOCAL_TEST_DIR}/test_esproducerlooper_prefer_producer_cfg.py ) || die 'Failure using test_esproducerlooper_prefer_producer_cfg.py' $?
 (cmsRun ${LOCAL_TEST_DIR}/test_module_change_looper_cfg.py ) || die 'Failure using test_module_change_looper_cfg.py' $?
-
+(cmsRun ${LOCAL_TEST_DIR}/test_edlooper_consumes_cfg.py) || die 'Failure using test_edlooper_consumes_cfg.py' $?

--- a/FWCore/Framework/test/stubs/IntTestLooper.cc
+++ b/FWCore/Framework/test/stubs/IntTestLooper.cc
@@ -1,0 +1,155 @@
+#include "DataFormats/TestObjects/interface/ToyProducts.h"
+#include "FWCore/Framework/interface/ESProducerLooper.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/LooperFactory.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Utilities/interface/transform.h"
+
+#include "FWCore/Framework/test/DummyData.h"
+#include "FWCore/Framework/test/DummyRecord.h"
+
+namespace edmtest {
+  class IntTestLooper : public edm::ESProducerLooper {
+  public:
+    IntTestLooper(edm::ParameterSet const& iPSet)
+        : tokensBeginRun_(edm::vector_transform(
+              iPSet.getUntrackedParameter<std::vector<edm::InputTag>>("srcBeginRun"),
+              [this](edm::InputTag const& tag) { return this->consumes<IntProduct, edm::InRun>(tag); })),
+          tokensBeginLumi_(edm::vector_transform(
+              iPSet.getUntrackedParameter<std::vector<edm::InputTag>>("srcBeginLumi"),
+              [this](edm::InputTag const& tag) { return this->consumes<IntProduct, edm::InLumi>(tag); })),
+          tokensEvent_(
+              edm::vector_transform(iPSet.getUntrackedParameter<std::vector<edm::InputTag>>("srcEvent"),
+                                    [this](edm::InputTag const& tag) { return this->consumes<IntProduct>(tag); })),
+          tokensEndLumi_(edm::vector_transform(
+              iPSet.getUntrackedParameter<std::vector<edm::InputTag>>("srcEndLumi"),
+              [this](edm::InputTag const& tag) { return this->consumes<IntProduct, edm::InLumi>(tag); })),
+          tokensEndRun_(edm::vector_transform(
+              iPSet.getUntrackedParameter<std::vector<edm::InputTag>>("srcEndRun"),
+              [this](edm::InputTag const& tag) { return this->consumes<IntProduct, edm::InRun>(tag); })),
+          esTokenBeginRun_(esConsumes<edm::Transition::BeginRun>()),
+          esToken_(esConsumes()),
+          valuesBeginRun_(iPSet.getUntrackedParameter<std::vector<int>>("expectBeginRunValues")),
+          valuesBeginLumi_(iPSet.getUntrackedParameter<std::vector<int>>("expectBeginLumiValues")),
+          valuesEvent_(iPSet.getUntrackedParameter<std::vector<int>>("expectEventValues")),
+          valuesEndLumi_(iPSet.getUntrackedParameter<std::vector<int>>("expectEndLumiValues")),
+          valuesEndRun_(iPSet.getUntrackedParameter<std::vector<int>>("expectEndRunValues")),
+          valueES_(iPSet.getUntrackedParameter<int>("expectESValue")) {}
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+      edm::ParameterSetDescription desc;
+      desc.setComment(
+          "This test EDLooper can consume arbitrary number of IntProduct's at Run, Lumi, and Event transitions. It "
+          "always consumes DummyData EventSetup product in beginRun and Event transitions. For all the consumed "
+          "products it requires the value of the product to correspond the value in configuratoin. It stops looping "
+          "after 3rd time.");
+      desc.addUntracked<std::vector<edm::InputTag>>("srcBeginRun", std::vector<edm::InputTag>{});
+      desc.addUntracked<std::vector<edm::InputTag>>("srcBeginLumi", std::vector<edm::InputTag>{});
+      desc.addUntracked<std::vector<edm::InputTag>>("srcEvent", std::vector<edm::InputTag>{});
+      desc.addUntracked<std::vector<edm::InputTag>>("srcEndLumi", std::vector<edm::InputTag>{});
+      desc.addUntracked<std::vector<edm::InputTag>>("srcEndRun", std::vector<edm::InputTag>{});
+      desc.addUntracked<std::vector<int>>("expectBeginRunValues", std::vector<int>{});
+      desc.addUntracked<std::vector<int>>("expectBeginLumiValues", std::vector<int>{});
+      desc.addUntracked<std::vector<int>>("expectEventValues", std::vector<int>{});
+      desc.addUntracked<std::vector<int>>("expectEndLumiValues", std::vector<int>{});
+      desc.addUntracked<std::vector<int>>("expectEndRunValues", std::vector<int>{});
+      desc.addUntracked<int>("expectESValue");
+      descriptions.addDefault(desc);
+    }
+
+    void beginRun(edm::Run const& iRun, edm::EventSetup const& iSetup) override {
+      for (size_t i = 0; i < tokensBeginRun_.size(); ++i) {
+        int const expectedValue = valuesBeginRun_[i];
+        int const value = iRun.get(tokensBeginRun_[i]).value;
+        if (expectedValue != value) {
+          throw cms::Exception("WrongValue")
+              << "expected value " << expectedValue << " but got " << value << " for beginRun product " << i;
+        }
+      }
+
+      auto const& data = iSetup.getData(esTokenBeginRun_);
+      if (data.value_ != valueES_) {
+        throw cms::Exception("WrongValue")
+            << " expected value " << valueES_ << " but got " << data.value_ << " for EventSetup product in beginRun";
+      }
+    }
+
+    void beginLuminosityBlock(edm::LuminosityBlock const& iLumi, edm::EventSetup const&) override {
+      for (size_t i = 0; i < tokensBeginLumi_.size(); ++i) {
+        int const expectedValue = valuesBeginLumi_[i];
+        int const value = iLumi.get(tokensBeginLumi_[i]).value;
+        if (expectedValue != value) {
+          throw cms::Exception("WrongValue")
+              << "expected value " << expectedValue << " but got " << value << " for beginLumi product " << i;
+        }
+      }
+    }
+
+    void endLuminosityBlock(edm::LuminosityBlock const& iLumi, edm::EventSetup const&) override {
+      for (size_t i = 0; i < tokensEndLumi_.size(); ++i) {
+        int const expectedValue = valuesEndLumi_[i];
+        int const value = iLumi.get(tokensEndLumi_[i]).value;
+        if (expectedValue != value) {
+          throw cms::Exception("WrongValue")
+              << "expected value " << expectedValue << " but got " << value << " for endLumi product " << i;
+        }
+      }
+    }
+
+    void endRun(edm::Run const& iRun, edm::EventSetup const&) override {
+      for (size_t i = 0; i < tokensEndRun_.size(); ++i) {
+        int const expectedValue = valuesEndRun_[i];
+        int const value = iRun.get(tokensEndRun_[i]).value;
+        if (expectedValue != value) {
+          throw cms::Exception("WrongValue")
+              << "expected value " << expectedValue << " but got " << value << " for endRun product " << i;
+        }
+      }
+    }
+
+    void startingNewLoop(unsigned int) override {}
+
+    Status duringLoop(edm::Event const& iEvent, edm::EventSetup const& iSetup) override {
+      for (size_t i = 0; i < tokensEvent_.size(); ++i) {
+        int const expectedValue = valuesEvent_[i];
+        int const value = iEvent.get(tokensEvent_[i]).value;
+        if (expectedValue != value) {
+          throw cms::Exception("WrongValue")
+              << "expected value " << expectedValue << " but got " << value << " for Event product " << i;
+        }
+      }
+
+      auto const& data = iSetup.getData(esToken_);
+      if (data.value_ != valueES_) {
+        throw cms::Exception("WrongValue")
+            << " expected value " << valueES_ << " but got " << data.value_ << " for EventSetup product in duringLoop";
+      }
+
+      return kContinue;
+    }
+
+    Status endOfLoop(edm::EventSetup const&, unsigned int iCount) override { return iCount == 2 ? kStop : kContinue; }
+
+  private:
+    const std::vector<edm::EDGetTokenT<IntProduct>> tokensBeginRun_;
+    const std::vector<edm::EDGetTokenT<IntProduct>> tokensBeginLumi_;
+    const std::vector<edm::EDGetTokenT<IntProduct>> tokensEvent_;
+    const std::vector<edm::EDGetTokenT<IntProduct>> tokensEndLumi_;
+    const std::vector<edm::EDGetTokenT<IntProduct>> tokensEndRun_;
+    edm::ESGetToken<edm::eventsetup::test::DummyData, edm::DefaultRecord> const esTokenBeginRun_;
+    edm::ESGetToken<edm::eventsetup::test::DummyData, edm::DefaultRecord> const esToken_;
+    const std::vector<int> valuesBeginRun_;
+    const std::vector<int> valuesBeginLumi_;
+    const std::vector<int> valuesEvent_;
+    const std::vector<int> valuesEndLumi_;
+    const std::vector<int> valuesEndRun_;
+    const int valueES_;
+  };
+}  // namespace edmtest
+
+using IntTestLooper = edmtest::IntTestLooper;
+DEFINE_FWK_LOOPER(IntTestLooper);

--- a/FWCore/Framework/test/stubs/TestModuleChangeLooper.cc
+++ b/FWCore/Framework/test/stubs/TestModuleChangeLooper.cc
@@ -16,7 +16,6 @@
 //
 
 // user include files
-#include "DataFormats/Common/interface/Handle.h"
 #include "DataFormats/TestObjects/interface/ToyProducts.h"
 #include "FWCore/Framework/interface/EDLooper.h"
 #include "FWCore/Framework/interface/Event.h"
@@ -38,21 +37,24 @@ public:
 
   void startingNewLoop(unsigned int) {}
   Status duringLoop(edm::Event const& iEvent, edm::EventSetup const&) {
-    edm::Handle<edmtest::IntProduct> handle;
-    iEvent.getByLabel(m_tag, handle);
-    if (handle->value != m_expectedValue) {
-      throw cms::Exception("WrongValue") << "expected value " << m_expectedValue << " but got " << handle->value;
+    auto const& product = iEvent.get(m_token);
+    if (product.value != m_expectedValue) {
+      throw cms::Exception("WrongValue") << "expected value " << m_expectedValue << " but got " << product.value;
     }
     return kContinue;
   }
   Status endOfLoop(edm::EventSetup const&, unsigned int iCount) {
     //modify the module
-    edm::ParameterSet const* pset = scheduleInfo()->parametersForModule(m_tag.label());
+    Labels labels;
+    labelsForToken(m_token, labels);
+    std::string const moduleLabel{labels.module};
+
+    edm::ParameterSet const* pset = scheduleInfo()->parametersForModule(moduleLabel);
     assert(0 != pset);
 
     edm::ParameterSet newPSet(*pset);
     newPSet.addParameter<int>("ivalue", ++m_expectedValue);
-    auto success = moduleChanger()->changeModule(m_tag.label(), newPSet);
+    auto success = moduleChanger()->changeModule(moduleLabel, newPSet);
     assert(success && "moduleChanger()->changeModule(m_tag.label(), newPSet)");
 
     return iCount == 2 ? kStop : kContinue;
@@ -61,7 +63,7 @@ public:
 private:
   // ----------member data ---------------------------
   int m_expectedValue;
-  edm::InputTag m_tag;
+  edm::EDGetTokenT<edmtest::IntProduct> m_token;
 };
 
 //
@@ -77,7 +79,7 @@ private:
 //
 TestModuleChangeLooper::TestModuleChangeLooper(edm::ParameterSet const& iConfig)
     : m_expectedValue(iConfig.getUntrackedParameter<int>("startingValue")),
-      m_tag(iConfig.getUntrackedParameter<edm::InputTag>("tag")) {
+      m_token(consumes(iConfig.getUntrackedParameter<edm::InputTag>("tag"))) {
   //now do what ever other initialization is needed
 }
 

--- a/FWCore/Framework/test/test_edlooper_consumes_cfg.py
+++ b/FWCore/Framework/test/test_edlooper_consumes_cfg.py
@@ -1,0 +1,29 @@
+import FWCore.ParameterSet.Config as cms
+
+process = cms.Process("TEST")
+
+process.maxEvents = cms.untracked.PSet(
+    input = cms.untracked.int32(5)
+)
+
+process.source = cms.Source("EmptySource")
+
+process.intEventProducer = cms.EDProducer("IntProducer",
+    ivalue = cms.int32(42)
+)
+
+process.essource = cms.ESSource("EmptyESSource",
+    recordName = cms.string('DummyRecord'),
+    iovIsRunNotTime = cms.bool(True),
+    firstValid = cms.vuint32(1)
+)
+process.add_(cms.ESProducer("LoadableDummyProvider",
+                            value = cms.untracked.int32(5)))
+
+process.looper = cms.Looper("IntTestLooper",
+    srcEvent = cms.untracked.VInputTag("intEventProducer"),
+    expectEventValues = cms.untracked.vint32(42),
+    expectESValue = cms.untracked.int32(5)
+)
+
+process.p1 = cms.Path(process.intEventProducer)

